### PR TITLE
iptables module: Allow passing additional arbitrary arguments to iptables rules

### LINKED ISSUE #72189: iptables module: allow use of additional arguments
##### SUMMARY

There are various extensions that can be used with iptables which have specific flags and options that are not explicitly mentioned in the main iptables CLI and not curently usable with the iptables module.

Example rule:
```
iptables -A FORWARD -j NFQUEUE --queue-num 0 --queue-bypass
```

Allowing additional arguments to be specified would be really useful. In the above example, `--queue-num 0 --queue-bypass` can't currently be used with the iptables module.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

iptables

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
##### SUMMARY

Fixes #72189

There are various extensions that can be used with iptables which have specific flags and options that are not explicitly mentioned in the main iptables CLI.

This patch adds support to the iptables module for supplying arbitrary additional arguments, for example to generate rules such as:

```
iptables -A FORWARD -j NFQUEUE --queue-num 0 --queue-bypass
```

This example rule is pretty common, but it's not currently possible to add/remove/manage it with the iptables module, as `--queue-num 0 --queue-bypass` is not supported.

There are lots of other extensions with too many options to list (for example: #68985), so I've implemented a generic solution that covers a broad range of potential use-cases.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

iptables

##### ADDITIONAL INFORMATION

Example usage:
```
- iptables:
    chain: FORWARD
    action: append
    jump: NFQUEUE
    extra_args: "--queue-num 0 --queue-bypass"
  become: yes
```
or:
```
- iptables:
    chain: FORWARD
    action: append
    jump: NFQUEUE
    extra_args:
      - "--queue-num 0"
      - "--queue-bypass"
  become: yes
```
will both create: 
```
iptables -A FORWARD -j NFQUEUE --queue-num 0 --queue-bypass
```

### CODE REVIEW

**Comment by Matt-Yorkley** [comment]
Also closes: #68985, #41147. Potentially closes: #58329, #69394.

File: lib/ansible/modules/iptables.py [hunk 1]
```diff
@@ -461,6 +467,15 @@
...
     log_level: info
+
+- name: Create rule with arbitrary additional arguments
+  iptables:
+    chain: INPUT
+    jump: NFQUEUE
+    extra_args:
+      - "--queue-num 0"
+      - "--queue-bypass"
+  become: yes
```
Comment (Akasurde) [inline lib/ansible/modules/iptables.py]:
```suggestion
```

File: lib/ansible/modules/iptables.py [hunk 2]
```diff
@@ -342,6 +342,12 @@
         the program from running concurrently.
     type: str
     version_added: "2.10"
+  extra_args:
+    description:
+      - List of additional arguments to be used with the rule.
+    type: list
+    elements: str
```
Comment (Akasurde) [inline lib/ansible/modules/iptables.py:349]:
```suggestion
    elements: str
    version_added: "2.11"
```

File: lib/ansible/modules/iptables.py [hunk 3]
```diff
@@ -342,6 +342,12 @@
         the program from running concurrently.
     type: str
     version_added: "2.10"
+  extra_args:
```
Comment (Akasurde) [inline lib/ansible/modules/iptables.py:345]: Add a changelog entry [here](https://github.com/ansible/ansible/tree/devel/changelogs/fragments)

**Comment by Akasurde** [comment]
@tsia @EvanDonato @sandorkazi @QuentinBrosse @mshkrebtan @smiller171 @r-daneel Could you please review this? Thanks.

**Review by sachudj — APPROVED** [review]
Looks Good

**Comment by sivel** [comment]
Question About This One
=======================

Hi!

Thanks very much for your submission to Ansible.  It sincerely means a lot to us that you've taken time to contribute.

Unfortunately, we're not sure if we want this feature in the program, and I don't want this to seem confrontational.  Our reasons for this are:

   * While this may be manageable for some small or simple extra args, adding support for generic `extra_args` often causes more issues than it solves, by allowing users to effectively break the module.
  * We would recommend creating new modules, or adding explicit options to this module to support explicit functionality.

However, we're absolutely always up for discussion.  Since this is a really busy project, we don't always see comments on closed tickets, but want to encourage
open dialog.  You can stop by the development list, and we'd be glad to talk about it - and we might even be persuaded otherwise!

   * https://groups.google.com/forum/#!forum/ansible-devel

In the future, sometimes starting a discussion on the development list prior to implementing a feature can make getting things included a little easier, but it's not always necessary.

Thank you once again for this and your interest in Ansible!

**Comment by Matt-Yorkley** [comment]
Hi @sivel, thanks for the response. I'll leave a quick reply here...

> starting a discussion on the development list prior to implementing a feature can make getting things included a little easier

If that's how the contributing process works in practice, it might be a good idea to explain that in the contributing guide :+1:

> adding support for generic extra_args often causes more issues than it solves, by allowing users to effectively break the module

I've tested it (beyond the specs in the PR) and it's working really well. Is the hypothetical problem here that the user could input bad or incorrect arguments which would cause an error? Isn't the same true of literally every Ansible module? If `iptables` really doesn't like the input, it will just give an error code and message and the task will fail, which is also normal module behavior, right?

> We would recommend creating new modules

Someone else already submitted a new module that would allow something like this (#36454), and it was rejected. To paraphrase, the responses and related discussions were basically; "we are no longer accepting new modules" and "why don't you incorporate this into the existing module?". A new module for this case would need all the same functionality of the existing module, so they would look almost identical.

> or adding explicit options to this module to support explicit functionality

I was intentionally trying to avoid that with this approach, as there are too many obscure options to count. Adding an explicit module arguments for every possible option seems like a poor solution? There are various other open issues with requests for obscure `iptables` flags to be added: #68985, #41147, #58329, #69394.

Basically the issue is that the `iptables` module is totally unusable with a wide assortment of perfectly reasonable iptables rules, and the only alternative is to write nasty tasks that interact with `iptables` directly via the `shell` module. Making them idempotent requires a mess of multiple tasks using `register:` and `when:`, etc.

**Comment by smiller171** [comment]
@Matt-Yorkley I tend to agree that for `iptables` specifically an `extra_args` flag might be good, but it's worth noting that Ansible considers any unhandled error that falls back to a stacktrace to be a bug (unless this has changed since the RedHat acquisition?), and it's not really possible to correctly handle all errors with an `extra_args` argument.

It may be the case that building a custom role using `shell` or `script` is the cleanest way forward for using more obscure iptables flags.

**Comment by Matt-Yorkley** [comment]
> it's not really possible to correctly handle all errors

We can test this hypothetical problem quite easily, and the result is a reasonable rejection response and a return code of 2:

```
user@example:$ sudo iptables -A INPUT -j ACCEPT alskdasdaskjkjncxc

Bad argument `alskdasdaskjkjncxc'
Try `iptables -h' or 'iptables --help' for more information.

user@example:$ echo $?

2
```

If this was done via the iptables module, the task would just result in a typical failure, with the respective error output available in `stderr_lines` and `rc` as normal.

**Comment by Matt-Yorkley** [comment]
```
user@example:$ sudo iptables -A INPUT -j ACCEPT what happens if I try doing this?
Bad argument `what'
Try `iptables -h' or 'iptables --help' for more information.

user@example:$ sudo iptables -A INPUT -j DROP --how --about --this?
iptables v1.8.4 (legacy): unknown option "--how"
Try `iptables -h' or 'iptables --help' for more information.

user@example:$ echo $?
2

user@example:$ sudo iptables -A FORWARD -j NFQUEUE --queue-num 0 --queue-bypass --OR --THIS?
iptables v1.8.4 (legacy): unknown option "--OR"
Try `iptables -h' or 'iptables --help' for more information.

user@example:$ echo $?
2

user@example:$ sudo iptables -A FORWARD -j NFQUEUE --queue-num aoisdu9984309
iptables v1.8.4 (legacy): NFQUEUE: bad value for option "--queue-num", or out of range (0-65535).
Try `iptables -h' or 'iptables --help' for more information.

user@example:$ echo $?
2

user@example:$ sudo iptables -L

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
```
:ok_hand:
